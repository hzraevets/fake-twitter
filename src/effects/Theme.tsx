import React, {
  Dispatch,
  SetStateAction,
  useEffect,
  useCallback,
  createContext,
  useState,
  ReactNode,
} from 'react';
import { ConfigProvider, theme as antdTheme } from "antd";

interface ThemeContextInterface {
  theme: string;
  setTheme: Dispatch<SetStateAction<string>>;
  toggleTheme: () => void;
}

const ThemeContext = createContext<ThemeContextInterface>({
  theme: 'light',
  setTheme: () => null,
  toggleTheme: () => null,
});

const getTheme = () => {
  const theme = localStorage.getItem('theme');
  if (!theme) { // default theme
    localStorage.setItem('theme', 'light');
    return 'light';
  }

  return theme;
};

const ThemeProvider = ({ children }: { children: ReactNode }) => {
  const [ theme, setTheme ] = useState(getTheme);

  function toggleTheme() {
    if (theme === 'dark') {
      setTheme('light');
    } else {
      setTheme('dark');
    }
  }

  const refreshTheme = useCallback(() => {
    localStorage.setItem('theme', theme);
  }, [theme]);

  useEffect(() => {
    refreshTheme();
  }, [refreshTheme, theme]);

  return (
    <ThemeContext.Provider
      value={{
        theme,
        setTheme,
        toggleTheme,
      }}
    >
      <div className={`theme-container ${theme}`}>
        <ConfigProvider
          theme={{
            algorithm: theme === 'dark'
              ? antdTheme.darkAlgorithm
              : antdTheme.defaultAlgorithm
          }}
        >
          {children}
        </ConfigProvider>
      </div>
    </ThemeContext.Provider>
  );
};

export { ThemeContext, ThemeProvider };
