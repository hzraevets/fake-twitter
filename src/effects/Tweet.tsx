import React, { createContext, useState, ReactNode } from 'react';
import getUnixTime from 'date-fns/getUnixTime';

import {
  Tweet,
  TweetStorage,
  CreateNewTweetParams,
  ReadNonDeleteTweetParams,
  ReadNonDeleteTweetResponse,
} from 'models';

export interface TweetContextInterface {
  allTweets: TweetStorage;
  nextTweetId: number;
  addNewTweet: (params: CreateNewTweetParams) => void;
  readTimeline: (params: ReadNonDeleteTweetParams) => Promise<ReadNonDeleteTweetResponse>;
  editTweet: (id: number, content: string) => void;
  deleteTweet: (id: number) => void;
}

const initialTweetStorage: TweetStorage = {
  hashMap: {},
  order: [],
};

const TweetContext = createContext<TweetContextInterface>({
  allTweets: initialTweetStorage,
  nextTweetId: 1,
  addNewTweet: () => null,
  readTimeline: () => Promise.resolve({ data: [] }),
  editTweet: () => null,
  deleteTweet: () => null,
});

const TweetProvider = ({ children }: { children: ReactNode }) => {
  const [ allTweets, setAllTweets ] = useState(initialTweetStorage);
  const [ nextTweetId, setNextTweetId ] = useState(1);

  function addNewTweet(params: CreateNewTweetParams) {
    const now = getUnixTime(new Date());
    const { hashMap, order } = allTweets;

    const newTweet: Tweet = {
      ...params,
      id: nextTweetId,
      createTime: now,
      updateTime: now,
      isDeleted: false,
    };

    setNextTweetId(nextTweetId + 1);
    setAllTweets({
      hashMap: { ...hashMap, [nextTweetId]: newTweet },
      order: [ nextTweetId, ...order ],
    });
  }

  function readTimeline({ start, offset = 1 }: ReadNonDeleteTweetParams) {
    const data: Tweet[] = [];
    let nextCursor: number | null;

    const { hashMap, order } = allTweets;

    for (nextCursor = start; (data.length < offset) && nextCursor < order.length; nextCursor++) {
      const tweet = hashMap[order[nextCursor]];

      if (!tweet.isDeleted) {
        data.push(tweet);
      }
    }

    const res: ReadNonDeleteTweetResponse = { data };

    if (nextCursor < order.length) {
      res.nextCursor = nextCursor;
    }

    return new Promise<ReadNonDeleteTweetResponse>(resolve =>
      setTimeout(() => resolve(res), Math.random() * 1000));
  }

  function editTweet(id: number, content: string) {
    const { hashMap, order } = allTweets;
    const oldTweet = hashMap[id];

    if (!oldTweet) {
      // This should not happen!
      return;
    }

    hashMap[id].content = content;
    hashMap[id].updateTime = getUnixTime(new Date());

    setAllTweets({ hashMap, order });
  }

  function deleteTweet(id: number) {
    const { hashMap } = allTweets;
    let order = allTweets.order;
    const oldTweet = hashMap[id];

    if (!oldTweet) {
      // This should not happen!
      return;
    }

    hashMap[id].isDeleted = true;
    order = order.filter(tid => tid !== id);

    setAllTweets({ hashMap, order });
  }

  return (
    <TweetContext.Provider
      value={{
        allTweets,
        nextTweetId,
        addNewTweet,
        readTimeline,
        editTweet,
        deleteTweet,
      }}
    >
      {children}
    </TweetContext.Provider>
  );
}

export { TweetContext, TweetProvider };
