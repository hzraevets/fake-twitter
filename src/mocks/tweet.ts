import {
  ReadNonDeleteTweetParams,
  ReadNonDeleteTweetResponse,
  Tweet,
  TweetStorage,
} from 'models';

export const testUserTweet1: Tweet = {
  id: 1,
  author: 'testUser',
  firstname: 'Test',
  content: 'Hello, world!',
  createTime: 1677073172,
  updateTime: 1677073172,
  isDeleted: false,
};

export const liBaiTweet1: Tweet = {
  id: 2,
  author: 'liBai',
  firstname: 'Li',
  content: '抽刀断水水更流，举杯销愁愁更愁',
  createTime: 1677250735,
  updateTime: 1677250775,
  isDeleted: false,
};

export const storage: TweetStorage = {
  hashMap: {
    [testUserTweet1.id]: testUserTweet1,
    [liBaiTweet1.id]: liBaiTweet1,
  },
  order: [
    liBaiTweet1.id,
    testUserTweet1.id,
  ],
};

export const nextTweetId = 3;

export const mockedReadTimeline = ({ start, offset = 1 }: ReadNonDeleteTweetParams) => {
  const dat: Tweet[] = [];
  let nextCursor: number | null;

  const { hashMap, order } = storage;

  for (nextCursor = start; (dat.length < offset) && nextCursor < order.length; nextCursor++) {
    const tweet = hashMap[order[nextCursor]];

    if (!tweet.isDeleted) {
      dat.push(tweet);
    }
  }

  const res: ReadNonDeleteTweetResponse = { data: dat };

  if (nextCursor < order.length) {
    res.nextCursor = nextCursor;
  }

  return Promise.resolve<ReadNonDeleteTweetResponse>(res);
}

