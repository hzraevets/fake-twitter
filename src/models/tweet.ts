export interface Tweet {
  id: number;
  author: string;
  firstname: string;
  content: string;
  createTime: number; // unix timestamp
  updateTime: number; // unix timestamp
  isDeleted: boolean;
}

export interface TweetStorage {
  hashMap: { [id: number]: Tweet };
  order: number[]; // desc for tweet according to create time
}

export type CreateNewTweetParams = Omit<
  Tweet,
  'id' | 'createTime' | 'updateTime' | 'isDeleted'
>;

export interface ReadNonDeleteTweetParams {
  start: number;
  offset?: number;
}

export interface ReadNonDeleteTweetResponse {
  data: Tweet[];
  nextCursor?: number | null;
}

export type UpdateTweetParams = Omit<
  Tweet,
  'firstname' | 'createTime' | 'updateTime' | 'isDeleted'
>;

export type DeleteTweetParams = Omit<
  Tweet,
  'content' | 'firstname' | 'createTime' | 'updateTime' | 'isDeleted'
>;
