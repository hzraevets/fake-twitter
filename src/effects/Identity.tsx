import React, {
  Dispatch,
  SetStateAction,
  useEffect,
  createContext,
  useState,
  ReactNode,
} from 'react';

import { UserStorage, User } from 'models';
import { md5 } from 'utils/md5';

export interface IdentityContextInterface {
  identity: UserStorage;
  setIdentity: Dispatch<SetStateAction<UserStorage>>;
  loginUser: string | null;
  setLoginUser: Dispatch<SetStateAction<string | null>>;
  logout: () => void;
  register: (newUser: User) => UserStorage;
}

const IdentityContext = createContext<IdentityContextInterface>({
  identity: {},
  setIdentity: () => null,
  loginUser: null,
  setLoginUser: () => null,
  logout: () => null,
  register: () => ({}),
});

const getIdentity = (): UserStorage => {
  const identity = localStorage.getItem('identity');
  if (!identity) {
    localStorage.setItem('identity', '{}');
    return {};
  }

  return JSON.parse(identity);
};

const getLoginUser = (): string | null => {
  const loginUser = localStorage.getItem('loginUser');
  if (!loginUser) {
    return null;
  }

  return loginUser;
};

const IdentityProvider = ({ children }: { children: ReactNode }) => {
  const [ identity, setIdentity ] = useState(getIdentity);
  const [ loginUser, setLoginUser ] = useState(getLoginUser);

  function logout() {
    if (loginUser) {
      setLoginUser(null);
      localStorage.removeItem('loginUser');
    }
  }

  function register(newUser: User) {
    const newStorage: UserStorage = {
      ...identity,
      [newUser.username]: {
        ...newUser,
        password: md5(newUser.password),
      },
    };

    localStorage.setItem('identity', JSON.stringify(newStorage));
    return newStorage;
  }

  useEffect(() => {
    const updateIdentity = () => {
      localStorage.setItem('identity', JSON.stringify(identity));
    };

    updateIdentity();
  }, [identity]);

  useEffect(() => {
    const executeLogin = () => {
      if (loginUser) {
        localStorage.setItem('loginUser', loginUser);
      }
    };

    executeLogin();
  }, [loginUser]);

  return (
    <IdentityContext.Provider
      value={{
        identity,
        setIdentity,
        loginUser,
        setLoginUser,
        logout,
        register,
      }}
    >
      {children}
    </IdentityContext.Provider>
  );
};

export { IdentityContext, IdentityProvider };
