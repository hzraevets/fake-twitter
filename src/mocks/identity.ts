import { User, UserStorage } from 'models';
import { md5 } from 'utils/md5';

export const testUser: User = {
  username: 'testUser',
  firstname: 'Test',
  password: 'abcde',
}

export const liBai: User = {
  username: 'liBai',
  firstname: 'Li',
  password: '12345',
};

export const storage: UserStorage = {
  testUser: { ...testUser, password: md5(testUser.password) },
  liBai: { ...liBai, password: md5(liBai.password) },
}
