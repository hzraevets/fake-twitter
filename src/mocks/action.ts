import { enableLocalStorageMock, enableMatchMedia } from 'utils/browserMocks';
import { TweetContextInterface } from 'effects/Tweet';
import {ReadNonDeleteTweetParams, ReadNonDeleteTweetResponse, Tweet} from 'models';
import { storage as identityStorage, testUser } from './identity';
import { nextTweetId, storage as tweetStorage } from './tweet';

export function prepareTweetProviderPropsWithoutCRUD(): TweetContextInterface {
  enableMatchMedia();
  enableLocalStorageMock();

  // prepare work for identity
  localStorage.setItem('identity', JSON.stringify(identityStorage));
  localStorage.setItem('loginUser', testUser.username);

  // prepare tweet provider props and return
  return {
    allTweets: tweetStorage,
    nextTweetId,
    addNewTweet: () => null,
    readTimeline: () => Promise.resolve({ data: [] }),
    editTweet: () => null,
    deleteTweet: () => null,
  };
}

export function prepareTweetProviderPropsWithRead(): TweetContextInterface {
  enableMatchMedia();
  enableLocalStorageMock();

  // prepare work for identity
  localStorage.setItem('identity', JSON.stringify(identityStorage));
  localStorage.setItem('loginUser', testUser.username);

  // prepare tweet provider props and return
  return {
    allTweets: tweetStorage,
    nextTweetId,
    addNewTweet: () => null,
    readTimeline: ({ start, offset = 1 }: ReadNonDeleteTweetParams) => {
      const dat: Tweet[] = [];
      let nextCursor: number | null;

      const { hashMap, order } = tweetStorage;

      for (nextCursor = start; (dat.length < offset) && nextCursor < order.length; nextCursor++) {
        const tweet = hashMap[order[nextCursor]];

        if (!tweet.isDeleted) {
          dat.push(tweet);
        }
      }

      const res: ReadNonDeleteTweetResponse = { data: dat };

      if (nextCursor < order.length) {
        res.nextCursor = nextCursor;
      }

      return Promise.resolve<ReadNonDeleteTweetResponse>(res);
    },
    editTweet: () => null,
    deleteTweet: () => null,
  };
}
