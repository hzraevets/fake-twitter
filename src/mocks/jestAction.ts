import { enableLocalStorageMock, enableMatchMedia } from 'utils/browserMocks';
import { TweetContextInterface } from 'effects/Tweet';
import { storage as identityStorage, testUser } from './identity';
import { nextTweetId, storage as tweetStorage, mockedReadTimeline } from './tweet';

export function prepareTweetProviderPropsWithoutCRUD(): TweetContextInterface {
  enableMatchMedia();
  enableLocalStorageMock();

  // prepare work for identity
  localStorage.setItem('identity', JSON.stringify(identityStorage));
  localStorage.setItem('loginUser', testUser.username);

  // prepare tweet provider props and return
  return {
    allTweets: tweetStorage,
    nextTweetId,
    addNewTweet: () => null,
    readTimeline: () => Promise.resolve({ data: [] }),
    editTweet: () => null,
    deleteTweet: () => null,
  };
}

export function prepareTweetProviderPropsWithRead(): TweetContextInterface {
  enableMatchMedia();
  enableLocalStorageMock();

  // prepare work for identity
  localStorage.setItem('identity', JSON.stringify(identityStorage));
  localStorage.setItem('loginUser', testUser.username);

  // prepare tweet provider props and return
  return {
    allTweets: tweetStorage,
    nextTweetId,
    addNewTweet: () => null,
    readTimeline: mockedReadTimeline,
    editTweet: () => null,
    deleteTweet: () => null,
  };
}
