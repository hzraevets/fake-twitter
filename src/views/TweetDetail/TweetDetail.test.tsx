import React, { ReactNode } from 'react';
import { render, screen } from '@testing-library/react';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
// @ts-ignore
import { create } from 'react-test-renderer';

import { TweetDetail } from './';
import { IdentityProvider } from 'effects/Identity';
import { TweetContextInterface, TweetContext } from 'effects/Tweet';
import { testUserTweet1, liBaiTweet1 } from 'mocks/tweet';
import { prepareTweetProviderPropsWithoutCRUD } from 'mocks/action';

const customRender = (
  tweetId: number,
  children: ReactNode,
  providerProps: TweetContextInterface,
) => render(
  <MemoryRouter initialEntries={[`/tweet/${tweetId}`]}>
    <Routes>
      <Route
        path="tweet/:id"
        element={(
          <TweetContext.Provider value={providerProps}>
            {children}
          </TweetContext.Provider>
        )}
      />
    </Routes>
  </MemoryRouter>
);

const snapshotCreate = (
  tweetId: number,
  children: ReactNode,
  providerProps: TweetContextInterface,
) => create(
  <MemoryRouter initialEntries={[`/tweet/${tweetId}`]}>
    <Routes>
      <Route
        path="tweet/:id"
        element={(
          <TweetContext.Provider value={providerProps}>
            {children}
          </TweetContext.Provider>
        )}
      />
    </Routes>
  </MemoryRouter>
);

describe('TweetDetail', () => {
  let providerProps: TweetContextInterface;

  beforeEach(() => {
    providerProps = prepareTweetProviderPropsWithoutCRUD();
  });

  it('render login user own tweet with TweetDetail page', () => {
    const { id, content, firstname } = testUserTweet1;

    customRender(id, (
      <IdentityProvider>
        <TweetDetail/>
      </IdentityProvider>
    ), providerProps);

    expect(screen.getByText(content)).toBeInTheDocument();
    expect(screen.getByText(firstname)).toBeInTheDocument();
    expect(screen.getByText(`Tweet id: ${id}`)).toBeInTheDocument();

    expect(screen.getByText('Edit')).toBeInTheDocument();
    expect(screen.getByText('Delete')).toBeInTheDocument();
  });

  it('render other user tweet with TweetDetail page', () => {
    const { id, content, firstname } = liBaiTweet1;

    customRender(liBaiTweet1.id, (
      <IdentityProvider>
        <TweetDetail/>
      </IdentityProvider>
    ), providerProps);

    expect(screen.getByText(content)).toBeInTheDocument();
    expect(screen.getByText(firstname)).toBeInTheDocument();
    expect(screen.getByText(`Tweet id: ${id}`)).toBeInTheDocument();

    expect(screen.queryByText('Edit')).toBeNull();
    expect(screen.queryByText('Delete')).toBeNull();
  });

  it('render login user own tweet to snapshot', () => {
    const { id } = testUserTweet1;
    const renderer = snapshotCreate(id, (
      <IdentityProvider>
        <TweetDetail/>
      </IdentityProvider>
    ), providerProps);
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('render other user tweet to snapshot', () => {
    const { id } = liBaiTweet1;
    const renderer = snapshotCreate(id, (
      <IdentityProvider>
        <TweetDetail/>
      </IdentityProvider>
    ), providerProps);
    expect(renderer.toJSON()).toMatchSnapshot();
  });
});
