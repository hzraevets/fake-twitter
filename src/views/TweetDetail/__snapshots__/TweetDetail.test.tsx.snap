// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TweetDetail render login user own tweet to snapshot 1`] = `
<div
  className="tweet-detail-container flex h-full"
>
  <div
    className="tweet-detail relative w-96 mx-auto h-fit pt-10"
  >
    <div
      className="ant-card ant-card-bordered pt-6 css-dev-only-do-not-override-diro6f"
    >
      <div
        className="ant-card-body"
        style={Object {}}
      >
        <a
          className="absolute top-1 left-2 cursor-pointer select-none"
          href="/testUser"
          onClick={[Function]}
        >
          <span
            className="text-2xl"
          >
            <span
              aria-label="arrow-left"
              className="anticon anticon-arrow-left"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="arrow-left"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M872 474H286.9l350.2-304c5.6-4.9 2.2-14-5.2-14h-88.5c-3.9 0-7.6 1.4-10.5 3.9L155 487.8a31.96 31.96 0 000 48.3L535.1 866c1.5 1.3 3.3 2 5.2 2h91.5c7.4 0 10.8-9.2 5.2-14L286.9 550H872c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
                />
              </svg>
            </span>
          </span>
        </a>
        <div
          className="absolute top-2 right-2 cursor-pointer select-none"
          onClick={[Function]}
        >
          <button
            aria-checked={false}
            className="ant-switch css-dev-only-do-not-override-diro6f"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="switch"
            type="button"
          >
            <div
              className="ant-switch-handle"
            />
            <span
              className="ant-switch-inner"
            >
              <span
                className="ant-switch-inner-checked"
              />
              <span
                className="ant-switch-inner-unchecked"
              />
            </span>
          </button>
        </div>
        <div
          className="ant-card-meta"
        >
          <div
            className="ant-card-meta-avatar"
          >
            <span
              className="ant-avatar ant-avatar-lg ant-avatar-circle ant-avatar-icon bg-slate-500 css-dev-only-do-not-override-diro6f"
              style={Object {}}
            >
              <span
                aria-label="user"
                className="anticon anticon-user"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="user"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M858.5 763.6a374 374 0 00-80.6-119.5 375.63 375.63 0 00-119.5-80.6c-.4-.2-.8-.3-1.2-.5C719.5 518 760 444.7 760 362c0-137-111-248-248-248S264 225 264 362c0 82.7 40.5 156 102.8 201.1-.4.2-.8.3-1.2.5-44.8 18.9-85 46-119.5 80.6a375.63 375.63 0 00-80.6 119.5A371.7 371.7 0 00136 901.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8 2-77.2 33-149.5 87.8-204.3 56.7-56.7 132-87.9 212.2-87.9s155.5 31.2 212.2 87.9C779 752.7 810 825 812 902.2c.1 4.4 3.6 7.8 8 7.8h60a8 8 0 008-8.2c-1-47.8-10.9-94.3-29.5-138.2zM512 534c-45.9 0-89.1-17.9-121.6-50.4S340 407.9 340 362c0-45.9 17.9-89.1 50.4-121.6S466.1 190 512 190s89.1 17.9 121.6 50.4S684 316.1 684 362c0 45.9-17.9 89.1-50.4 121.6S557.9 534 512 534z"
                  />
                </svg>
              </span>
            </span>
          </div>
          <div
            className="ant-card-meta-detail"
          >
            <div
              className="ant-card-meta-title"
            >
              <span
                className="text-sm font-bold leading-none select-none"
              >
                Test
              </span>
              <span
                className="text-xs text-slate-500 float-right select-none mt-1"
              >
                created 
                3 days ago
              </span>
            </div>
            <div
              className="ant-card-meta-description"
            >
              <p
                className="m-0 break-words text-sm"
              >
                Hello, world!
              </p>
              <div
                className="extra-info mt-4"
              >
                <span
                  className="tweet-id text-xs text-slate-600 font-bold select-none"
                >
                  Tweet id: 
                  1
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <ul
        className="ant-card-actions"
      >
        <li
          style={
            Object {
              "width": "50%",
            }
          }
        >
          <span>
            <span
              className="hover:text-blue-600 select-none inline-block w-full"
              onClick={[Function]}
            >
              Edit
            </span>
          </span>
        </li>
        <li
          style={
            Object {
              "width": "50%",
            }
          }
        >
          <span>
            <span
              className="hover:text-red-600 select-none inline-block w-full"
              onClick={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              onTouchStart={[Function]}
            >
              Delete
            </span>
          </span>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`TweetDetail render other user tweet to snapshot 1`] = `
<div
  className="tweet-detail-container flex h-full"
>
  <div
    className="tweet-detail relative w-96 mx-auto h-fit pt-10"
  >
    <div
      className="ant-card ant-card-bordered pt-6 css-dev-only-do-not-override-diro6f"
    >
      <div
        className="ant-card-body"
        style={Object {}}
      >
        <a
          className="absolute top-1 left-2 cursor-pointer select-none"
          href="/testUser"
          onClick={[Function]}
        >
          <span
            className="text-2xl"
          >
            <span
              aria-label="arrow-left"
              className="anticon anticon-arrow-left"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="arrow-left"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M872 474H286.9l350.2-304c5.6-4.9 2.2-14-5.2-14h-88.5c-3.9 0-7.6 1.4-10.5 3.9L155 487.8a31.96 31.96 0 000 48.3L535.1 866c1.5 1.3 3.3 2 5.2 2h91.5c7.4 0 10.8-9.2 5.2-14L286.9 550H872c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
                />
              </svg>
            </span>
          </span>
        </a>
        <div
          className="absolute top-2 right-2 cursor-pointer select-none"
          onClick={[Function]}
        >
          <button
            aria-checked={false}
            className="ant-switch css-dev-only-do-not-override-diro6f"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="switch"
            type="button"
          >
            <div
              className="ant-switch-handle"
            />
            <span
              className="ant-switch-inner"
            >
              <span
                className="ant-switch-inner-checked"
              />
              <span
                className="ant-switch-inner-unchecked"
              />
            </span>
          </button>
        </div>
        <div
          className="ant-card-meta"
        >
          <div
            className="ant-card-meta-avatar"
          >
            <span
              className="ant-avatar ant-avatar-lg ant-avatar-circle ant-avatar-icon bg-slate-500 css-dev-only-do-not-override-diro6f"
              style={Object {}}
            >
              <span
                aria-label="user"
                className="anticon anticon-user"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="user"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M858.5 763.6a374 374 0 00-80.6-119.5 375.63 375.63 0 00-119.5-80.6c-.4-.2-.8-.3-1.2-.5C719.5 518 760 444.7 760 362c0-137-111-248-248-248S264 225 264 362c0 82.7 40.5 156 102.8 201.1-.4.2-.8.3-1.2.5-44.8 18.9-85 46-119.5 80.6a375.63 375.63 0 00-80.6 119.5A371.7 371.7 0 00136 901.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8 2-77.2 33-149.5 87.8-204.3 56.7-56.7 132-87.9 212.2-87.9s155.5 31.2 212.2 87.9C779 752.7 810 825 812 902.2c.1 4.4 3.6 7.8 8 7.8h60a8 8 0 008-8.2c-1-47.8-10.9-94.3-29.5-138.2zM512 534c-45.9 0-89.1-17.9-121.6-50.4S340 407.9 340 362c0-45.9 17.9-89.1 50.4-121.6S466.1 190 512 190s89.1 17.9 121.6 50.4S684 316.1 684 362c0 45.9-17.9 89.1-50.4 121.6S557.9 534 512 534z"
                  />
                </svg>
              </span>
            </span>
          </div>
          <div
            className="ant-card-meta-detail"
          >
            <div
              className="ant-card-meta-title"
            >
              <span
                className="text-sm font-bold leading-none select-none"
              >
                Li
              </span>
              <span
                className="text-xs text-slate-500 float-right select-none mt-1"
              >
                created 
                about 20 hours ago
              </span>
            </div>
            <div
              className="ant-card-meta-description"
            >
              <p
                className="m-0 break-words text-sm"
              >
                抽刀断水水更流，举杯销愁愁更愁
              </p>
              <div
                className="extra-info mt-4"
              >
                <span
                  className="tweet-id text-xs text-slate-600 font-bold select-none"
                >
                  Tweet id: 
                  2
                </span>
                <span
                  className="update-time text-xs text-slate-500 float-right select-none"
                >
                  edited 
                  about 20 hours ago
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
